### PR TITLE
Added damaging through Veins.

### PR DESCRIPTION
**Pull request by Mailaender**
`-PoisonedByTiberium:` is currently the http://modenc.renegadeprojects.com/ImmuneToVeins equivalent. This is the easy part of getting veins to work.

### CODE REVIEW

File: mods/ts/rules/defaults.yaml [hunk 1]
```diff
@@ -417,6 +417,8 @@
 		Weapons: SmallDebris
 		Pieces: 3, 7
 		Range: 2c0, 5c0
+	PoisonedByTiberium:
+		Resources: Veins
```
Comment (pchote) [inline mods/ts/rules/defaults.yaml]: This will damage vehicles with the "Tiberium" weapon, which probably doesn't make sense.

Comment (pchote) [inline mods/ts/rules/defaults.yaml]: Rules.ini defines `VeinDamage=5`.

Comment (Mailaender) [inline mods/ts/rules/defaults.yaml]: I have no idea what `VeinDamage=5` has to be translated to. The default damage makes sense. It damages vehicles slowly.

Comment (GraionDilach) [inline mods/ts/rules/defaults.yaml]: Sounds like 5 damage per tick to me, `VeinDamage` is.

Comment (pchote) [inline mods/ts/rules/defaults.yaml]: The question is per *whose* tick?  (per vein grab, per second, or per TS game tick?)

Comment (pchote) [inline mods/ts/rules/defaults.yaml]: You can/should also set up the vein grab animation as a weapon effect.

Comment (GraionDilach) [inline mods/ts/rules/defaults.yaml]: I haven't tested it but I'd assume TS game tick.

**Comment by Mailaender** [comment]
Adjusted the vein damage and setup the vein tentacle sprites.

**Comment by reaperrr** [comment]
- The Nod bike needs to be made immune to them as well
- the tentacle animation is only displayed very rarely, but it should loop constantly while it does damage

**Comment by LipkeGu** [comment]
@reaperrr  Nod bikes arent veine immune! keep this in mind ... :)

**Comment by Mailaender** [comment]
I am not sure why the tentacle animation is only displayed when the vehicle moves, because damage is dealt constantly. Added support for infantry avoiding Tiberium patches and vehicles avoiding vein pits.

File: OpenRA.Mods.Cnc/Traits/PoisonedByTiberium.cs [hunk 2]
```diff
@@ -49,7 +49,7 @@ public void Tick(Actor self)
 			if (r == null || !Info.Resources.Contains(r.Info.Name))
 				return;
 
-			Info.WeaponInfo.Impact(Target.FromActor(self), self.World.WorldActor, Enumerable.Empty<int>());
+			Info.WeaponInfo.Impact(Target.FromPos(self.CenterPosition), self.World.WorldActor, Enumerable.Empty<int>());
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/PoisonedByTiberium.cs]: Why change this? The old code is more logically correct, so if that doesn't work for some reason then something else will need to be fixed.

Comment (Mailaender) [inline OpenRA.Mods.Cnc/Traits/PoisonedByTiberium.cs]: It crashes with the exceptions I mentioned in https://github.com/OpenRA/OpenRA/pull/10813/commits/7b80114e9c6e3cfcbe71a3c4dd1df5026e40f975.

Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/PoisonedByTiberium.cs]: That will only happen if the actor is already dead, so you can check and bail out inside DoImpact instead.

File: mods/ts/weapons/otherweapons.yaml [hunk 3]
```diff
@@ -66,3 +66,14 @@ Tiberium:
...
+Veins:
+	ReloadDelay: 16
+	Warhead@Damage: SpreadDamage
+		Spread: 42
+		Damage: 5
+		DamageTypes: EnergyDeath # TODO: FIX ME!
+	Warhead@Effect: CreateEffect
+		Explosions: veins
+		ExplosionPalette: player
+		ValidImpactTypes: Ground
```
Comment (reaperrr) [inline mods/ts/weapons/otherweapons.yaml]: This makes it display ONLY if the target is 'missed'. You need to add `GroundHit` as well.

Comment (GraionDilach) [inline mods/ts/weapons/otherweapons.yaml]: Or remove it.

**Comment by Mailaender** [comment]
Thanks for the help. Should be fixed now.

**Comment by reaperrr** [comment]
:+1: 

@OpenRA/engine-hackers 
I'm wondering whether the last commit that checks for invalid targets in `CreateEffectWarhead.DoImpact` should actually go into prep/stable once this is merged? It fixes a potential crash, afterall.

**Comment by Mailaender** [comment]
You can just cherry-pick https://github.com/OpenRA/OpenRA/pull/10813/commits/133a0f54a3673a84ebfff671fc967421d81af3f2 if you want it in prep/stable.

**Comment by reaperrr** [comment]
Yes, assuming someone else reviews and merges this quickly enough.

**Comment by GraionDilach** [comment]
This looks good as-is. :+1:

I'm not sure about the cherry-pick though. It's a partial fix to the issue #10091 brings up and I'm unsure how to treat it.

**Comment by GraionDilach** [comment]
Meh, I reconsidered, I'm :+1:ing the cherry-pick idea as well. The other warheads are already aware of the edge-case #10091 brings up (this includes `CreateEffectWarhead` as well) so indeed it makes sense.

**Pull request merged**